### DISCUSSION TITLE
How to find out which button is clicked when creating multiple buttons in loop?

### DISCUSSION THREAD

**Discussion by Anindya088** [zauberzeug/nicegui#1169] [Q&A] [answered]
### Question

Hi All,

I am importing an excel file, and then creating buttons for each sheet. Idea is that they will click the button and then they can see the sheet data. 

I can do this for a single button, by manually specifying the sheet name. Ex: like this
```python
data = pd.read_excel('test_excel.xlsm', sheet_name=None)
sh = list(data.keys())[4]     # Taking a random sheet for example
ui.button(sh,on_click=lambda: show_data())

def show_data():
    df = media_data[sh]
    ui.table(
        columns=[{'name': col, 'label': col, 'field': col} for col in df.columns],
        rows=df.to_dict('records'),
    )
```

But how do I convert this into a loop and do it for all sheets?

I tried something like this below:
```python
with ui.row():
    for sh, df in data.items():
       ui.button(sh, on_click=lambda: show_data())
```
But the problem I understood is that because this is a loop, the final value of ```sh``` is always the last sheet name.
So when I click some button, it's always the last sheet data 😞 . Although the button text is correctly done. 

How do I write the ```on_click``` function such that it knows which button is clicked, or find out the button text and pass it to the function such that I can display that specific sheet?

Kindly let me know what would be right way to approach this please. Many thanks for all the help.

Thanks,
Anindya

**Comment by akai-1024** [comment]
You'd better put the parameter in callback. Like this:

```python

table = ui.table(rows=[], columns=[])

def show_data(sh):
    df = media_data[sh]
    table.rows[:] = df.to_dict('records')
    table.update()

with ui.row():
    for sh, df in data.items():
       ui.button(sh, on_click=lambda: show_data(sh))
```

**Reply by falkoschindler** [reply, marked as answer]
Apart from that there's an issue with Python's "late binding":

Try
```py
for i in [1, 2, 3]:
    ui.button(i, on_click=lambda: ui.label(i))
```

vs.
```py
for i in [1, 2, 3]:
    ui.button(i, on_click=lambda i=i: ui.label(i))
```

The `i=i` captures the `i` within the lambda statement. When the lambda is eventually evaluated, it would use the current value of `i` (which is now 3). But with `i=i` the label is created using a local copy.

**Reply by Anindya088** [reply]
Hi @akai-1024 , @falkoschindler , 

Thanks to both of you, this is very helpful. I managed to solve the problem using ```i=i``` inside the lambda function as you suggested. 

@akai-1024 , one question. I saw you created the table above and then in the function you are updating the rows based on the data frame and the sheet selected. However since the columns are not defined yet, so the table is coming blank. 

Is there a way to update the columns through the function like you did using ```table.rows``` ? So my columns will be basically 
```python
df = data[sh]
columns=[{'name': col, 'label': col, 'field': col} for col in df.columns],
``` 
But if I have declared the table above, how can I now update the columns in this table now using the function?

So for now what I did was declaring a blank container above and then clearing it every time. That way I am creating the ```ui.table()``` inside the function every time and deleting it again when you press it next time.

But if you could tell me how I can update columns of the already created table that would be great :)

Here's what I have done so far:
```python
data = pd.read_excel('test_excel.xlsm', sheet_name=None)

with ui.row():
    for sh, df in data.items():
       ui.button(sh, on_click=lambda sh=sh: show_data(sh))

container = ui.row()

def show_data(sh):
    df = data[sh]
    container.clear()
    with container:
        ui.table(
            columns=[{'name': col, 'label': col, 'field': col} for col in df.columns],
            rows=df.to_dict('records'),
        )
```

**Reply by Baytars** [reply]
Thank you, you made my day.

**Reply by me21** [reply]
For future reference: it also works for me using https://docs.python.org/3/library/functools.html#functools.partial to bind the parameter value. Maybe someone would find it cleaner than lambda's default argument.
